lib/ok.ex

```diff
@@ -83,6 +83,38 @@ defmodule OK do
     if is_list(result), do: {:ok, Enum.reverse(result)}, else: result
   end
 
+  @doc """
+  Takes a result tuple, a predicate function, and an optional error reason.
+  If the result tuple is tagged as a success then its value will be passed to the predicate function.
+  If the predicate returns `true`, then the result tuple stay the same.
+  If the predicate returns `false, then the result tuple becomes `{:error, reason}`.
+  If the tag is failure then the predicate function is skipped.
+
+  ## Examples
+
+      iex> OK.filter({:ok, 2}, fn (x) -> x == 2 end)
+      {:ok, 2}
+
+      iex> OK.filter({:ok, 2}, fn (x) -> x == 3 end, :bad_value)
+      {:error, :bad_value}
+
+      iex> OK.filter({:error, :some_reason}, fn (x) -> x == 4 end)
+      {:error, :some_reason}
+  """
+  @spec filter({:ok, a} | {:error, reason}, (a -> boolean), test_failure_reason) ::
```

Comment:
Can you write this as a multi headed spec.

```
@spec filter({:error, reason}, (any ->boolean), any) :: {:error reason}
# and success match
```
This allows dialiser to realise that if it knows a value is already an error then it will never return an ok.
```

Reply:
Didn't know it was possible (as I said, I'm new to Elixir :wink:), I'll do that !

Reply:
Welcome to elixir (as I forgot to say earlier), the waters lovely :-)